test/unit/knex.js

```diff
@@ -468,4 +469,37 @@ describe('knex', () => {
         });
     });
   });
+
+  describe('extend query builder', () => {
+    let connection;
+    beforeEach(() => {
+      connection = new sqlite3.Database(':memory:');
+    });
+
+    afterEach(() => {
+      connection.close();
+      delete QueryBuilder.prototype.customSelect;
+    });
+
+    it('should extend default queryBuilder', (done) => {
+      Knex.QueryBuilder.extend('customSelect', function(value) {
+        return this.select(this.client.raw(`${value} as value`));
+      });
+
+      const knex = Knex({ client: 'sqlite3' });
+      knex
+        .connection(connection)
+        .customSelect(42)
+        .then((result) => {
+          expect(result[0].value).to.equal(42);
+          done();
+        });
+    });
+
```

Comment:
There should be also tests that custom method is available for example in transaction and maybe in some cases one likes to add it to join builder or some nested subquery builder.

Reply:
Not sure I'm following, the extend is done on the Builder prototype only once in the app.
There is no difference if you are using the custom method with sub-query or any other chain.

I'm missing something?

Reply:
Transaction is a different entity, not an original knex builder, but user would be justified to expect custom methods on it as well. It is highly likely we'd need to copy them over, but test would show that. Same with join builder, it's a different thing.

Reply:
So, should I add `extend` to them as well? should it be separate method for each?

Reply:
No, it would be better if they were inherited automatically. User needs not to concern himself with difference between all three.

Reply:
Agree, do Transaction & Join Builder are inherites from Builder? If so, the requirement is to check that it extends the prototype, meaning that is available at the child class?

Reply:
Nope:
src/transaction.js
src/query/joinclause.js (TBH, join builder is a farily different beast, I'm not sure querybuilder methods make much sense there. @elhigu, what use-case did you have in mind?)

Reply:
I'm not sure what is going to be safest and cleanest solution here, TBH. Maybe QueryBuilder.extend could also extend prototype of Transaction, or overall .extend method could be moved to general Knex.extend that covers both QueryBuilder and Transaction.

Reply:
What is the use case of extending join & transaction?

Reply:
You can use transaction pretty much same way you can use knex instance. See example from documentation:
```
// Using trx as a transaction object:
const trx = await knex.transaction();

const books = [
  {title: 'Canterbury Tales'},
  {title: 'Moby Dick'},
  {title: 'Hamlet'}
];

trx('catalogues')
  .insert({name: 'Old Books'}, 'id')
  .then(function(ids) {
    books.forEach((book) => book.catalogue_id = ids[0]);
    return trx('books').insert(books);
  })
  .then(trx.commit)
  .catch(trx.rollback);
})
```

Considering that currently transaction supports same query-building methods that knex.QueryBuilder does, as a user I would be very surprised if I'm supposed to set those up separately.

Reply:
Although looking at the code, I suspect this is going to just work (test to prove that would be most welcome, though). The way transactor is created is this:
```
  const transactor = makeKnex(trxClient);
```
So what gets called is this:
```
function makeKnex(client) {
  // The object we're potentially using to kick off an initial chain.
  function knex(tableName, options) {
    return createQueryBuilder(knex.context, tableName, options);
  }

  redefineProperties(knex, client);
  return knex;
}
```

Unless I'm missing something, this is going to call same QueryBuilder constructor in the end.

Reply:
Ha, I'm not using transaction in that way ;)
```javascript
knex.transaction((trx) => {
  knex(t1).transacting(trx)...
})
```
If it has the same query methods, it make sense to extend it as well.

Reply:
`trx` var in this context is the same transactor you would get from `await knex.transaction()`, and see above, I suspect it already works.

Reply:
Let me add a test for it to assure it, if it not works we will need to think how extend it as well